src/modules/utils.js

```diff
@@ -27,17 +37,119 @@ export const useDidUpdateEffect = (fn, inputs) => {
     }, inputs)
 }
 
-export const formatDimensionId = (dimensionId, programStageId) =>
-    programStageId ? `${programStageId}.${dimensionId}` : dimensionId
+export const formatDimensionId = ({
+    dimensionId,
+    programStageId,
+    programId,
+    outputType,
+}) => {
+    return [
+        outputType === OUTPUT_TYPE_TRACKED_ENTITY ? programId : undefined,
+        programStageId,
+        dimensionId,
+    ]
+        .filter((p) => p)
+        .join('.')
+}
 
-export const extractDimensionIdParts = (input) => {
-    const [dimensionId, rawStageId] = input.split('.').reverse()
+export const extractDimensionIdParts = (id, inputType) => {
+    let rawStageId
+    const [dimensionId, part2, part3] = id.split('.').reverse()
+    let programId = part3
+    if (part3 || inputType !== OUTPUT_TYPE_TRACKED_ENTITY) {
+        rawStageId = part2
+    }
+    if (inputType === OUTPUT_TYPE_TRACKED_ENTITY && !part3) {
+        programId = part2
+    }
     const [programStageId, repetitionIndex] = (rawStageId || '').split('[')
     return {
         dimensionId,
         programStageId,
+        ...(programId ? { programId } : {}),
         repetitionIndex:
             repetitionIndex?.length &&
             repetitionIndex.substring(0, repetitionIndex.indexOf(']')),
     }
 }
+
+export const getDimensionsWithSuffix = ({
+    dimensionIds,
+    metadata,
+    inputType,
+}) => {
+    const dimensions = dimensionIds.map((id) => {
+        const { dimensionId, programStageId, programId } =
+            extractDimensionIdParts(id, inputType)
+        const dimension = {
+            ...metadata[id],
+            dimensionId,
+            programStageId,
+            programId,
+        }
+        if (!dimension.id) {
+            dimension.id = id
+        }
+        return dimension
+    })
+
+    if (
+        [OUTPUT_TYPE_ENROLLMENT, OUTPUT_TYPE_TRACKED_ENTITY].includes(inputType)
```

Comment:
I agree that checking for `EVENT` would be a lot smoother. But as I added the TE type I realised that we had a lot of implicit targeting (like `if(!inputType ===  OUTPUT_TYPE_EVENT)` to target enrollment implicitly), which would then apply to TE implicitly as well (with mixed results). So instead I've tried to target explicitly instead, to avoid future mistakes. But since there's such a clear separation between the two paths here, maybe a good compromise would be to do 
```
if (![OUTPUT_TYPE_ENROLLMENT, OUTPUT_TYPE_TRACKED_ENTITY].includes(inputType)) {
    return dimensions
}
```